src/lib/x509/certstor_system_macos/certstor_macos.cpp

```diff
@@ -0,0 +1,475 @@
+      throw Lookup_Error("ambiguous certificate result");
+      }
+
+   return certs.front();
+   }
+
+std::vector<std::shared_ptr<const X509_Certificate>> Certificate_Store_MacOS::find_all_certs(
+         const X509_DN& subject_dn,
+         const std::vector<uint8_t>& key_id) const
+   {
+   std::vector<uint8_t> dn_data;
+   DER_Encoder encoder(dn_data);
+   normalize(subject_dn).encode_into(encoder);
+
+   scoped_CFType<CFDataRef> dn_cfdata(createCFDataView(dn_data));
+   check_notnull(dn_cfdata, "create DN search object");
+
+   Certificate_Store_MacOS_Impl::Query query_params(
+      {
+         {kSecAttrSubject, dn_cfdata.get()}
+      });
+
+   scoped_CFType<CFDataRef> keyid_cfdata(createCFDataView(key_id));
+   check_notnull(keyid_cfdata, "create key ID search object");
+   if(!key_id.empty())
+      {
+      query_params.push_back({kSecAttrSubjectKeyID, keyid_cfdata.get()});
+      }
+
+   scoped_CFType<CFArrayRef> result(m_impl->search(std::move(query_params)));
+
+   if(!result)
+      {
+      return {};  // no certificates found
+      }
+
+   const auto count = CFArrayGetCount(result.get());
+   BOTAN_ASSERT(count > 0, "certificate result list contains data");
+
+   std::vector<std::shared_ptr<const X509_Certificate>> output;
+   output.reserve(count);
+   for(unsigned int i = 0; i < count; ++i)
+      {
+      auto cfCert = to_SecCertificateRef(CFArrayGetValueAtIndex(result.get(), i));
+      output.emplace_back(readCertificate(cfCert));
+      }
+
+   return output;
+   }
+
+std::shared_ptr<const X509_Certificate>
+Certificate_Store_MacOS::find_cert_by_pubkey_sha1(const std::vector<uint8_t>& key_hash) const
+   {
+   if(key_hash.size() != 20)
+      {
+      throw Invalid_Argument("Certificate_Store_MacOS::find_cert_by_pubkey_sha1 invalid hash");
+      }
+
+   scoped_CFType<CFDataRef> key_hash_cfdata(createCFDataView(key_hash));
+   check_notnull(key_hash_cfdata, "create key hash search object");
+
+   scoped_CFType<CFArrayRef> result(m_impl->search(
+      {
+         {kSecAttrPublicKeyHash, key_hash_cfdata.get()},
+      }));
+
+   if(!result)
+      {
+      return nullptr;  // no certificate found
+      }
+
+   const auto count = CFArrayGetCount(result.get());
+   BOTAN_ASSERT(count > 0, "certificate result list contains an object");
+
+   // `count` might be greater than 1, but we'll just select the first match
+   auto cfCert = to_SecCertificateRef(CFArrayGetValueAtIndex(result.get(), 0));
+   return readCertificate(cfCert);
+   }
+
+std::shared_ptr<const X509_Certificate>
+Certificate_Store_MacOS::find_cert_by_raw_subject_dn_sha256(const std::vector<uint8_t>& subject_hash) const
+   {
+   BOTAN_UNUSED(subject_hash);
+   throw Not_Implemented("Certificate_Store_MacOS::find_cert_by_raw_subject_dn_sha256");
+   }
+
+std::shared_ptr<const X509_CRL> Certificate_Store_MacOS::find_crl_for(const X509_Certificate& subject) const
+   {
+   BOTAN_UNUSED(subject);
+   throw Not_Implemented("Certificate_Store_MacOS::find_crl_for");
```

Comment:
I'm currently testing this along with our under-construction TLS asio stream (see #1731) and it works quite nicely. However, for the certificate validation to work, I need to statically return an empty list of CRLs (or respectively don't override this function at all, hence falling back to the base class' method returning an empty list). As far as I can see, the macOS keychain does not support any CRL handling. Is "not overriding" a reasonable implementation strategy here?

Reply:
Yes if the keychain does not support CRLs at all then it is fine to either omit implementing the function entirely, or (to make it obvious the omission is intentional) override it and have it return an empty list with a comment to the effect that it is not supported by the system API. `Not_Implemented` is more to indicate a TODO type situation, where the functionality is possible but the code is just not written yet.